client/app/components/QuerySelector.jsx

```diff
@@ -0,0 +1,160 @@
+
+class StaleSearchError extends Error {
+  constructor() {
+    super('stale search');
+  }
+}
+
+function search(term) {
+  // get recent
+  if (!term) {
+    return Query.recent().$promise
+      .then((results) => {
+        const filteredResults = results.filter(item => !item.is_draft); // filter out draft
+        return Promise.resolve(filteredResults);
+      });
+  }
+
+  // search by query
+  return Query.query({ q: term }).$promise
+    .then(({ results }) => Promise.resolve(results));
+}
+
+export function QuerySelector(props) {
+  const [searchTerm, setSearchTerm] = useState();
+  const [searching, setSearching] = useState();
+  const [searchResults, setSearchResults] = useState([]);
+  const [selectedQuery, setSelectedQuery] = useState();
+
+  let isStaleSearch = false;
+  const debouncedSearch = debounce(_search, SEARCH_DEBOUNCE_DURATION);
+  const placeholder = 'Search a query by name';
+  const clearIcon = <i className="fa fa-times" onClick={() => selectQuery(null)} />;
+  const spinIcon = <i className={cx('fa fa-spinner fa-pulse', { hidden: !searching })} />;
+
+  // set selected from prop
+  useEffect(() => {
+    if (props.selectedQuery) {
+      setSelectedQuery(props.selectedQuery);
+    }
+  }, [props.selectedQuery]);
+
+  // on search term changed, debounced
+  useEffect(() => {
+    // clear results, no search
+    if (searchTerm === null) {
+      setSearchResults(null);
+      return () => {};
+    }
+
+    // search
+    debouncedSearch(searchTerm);
+    return () => {
+      debouncedSearch.cancel();
+      isStaleSearch = true;
+    };
+  }, [searchTerm]);
+
+  function _search(term) {
+    setSearching(true);
+    search(term)
+      .then(rejectStale)
+      .then((results) => {
+        setSearchResults(results);
+        setSearching(false);
+      })
+      .catch((err) => {
+        if (!(err instanceof StaleSearchError)) {
+          setSearching(false);
+        }
+      });
+  }
+
+  function rejectStale(results) {
+    return isStaleSearch
+      ? Promise.reject(new StaleSearchError())
+      : Promise.resolve(results);
+  }
+
+  function selectQuery(queryId) {
+    let query = null;
+    if (queryId) {
+      query = find(searchResults, { id: queryId });
+      if (!query) { // shouldn't happen
+        toastr.error('Something went wrong... Couldn\'t select query');
+      }
+    }
+
+    setSearchTerm(query ? null : ''); // empty string triggers recent fetch
+    setSelectedQuery(query);
+    props.onChange(query);
+  }
+
+  function renderResults() {
+    if (!searchResults.length) {
+      return <div className="text-muted">No results matching search term.</div>;
+    }
+
+    return (
+      <div className="list-group">
+        {searchResults.map(q => (
+          <a
+            href="javascript:void(0)"
+            className={cx('list-group-item', { inactive: q.is_draft })}
+            key={q.id}
+            onClick={() => selectQuery(q.id)}
+          >
+            {q.name}
```

Comment:
BTW - how about highlighting search phrase (https://github.com/getredash/redash/pull/3494/files#diff-9487bf74e60479e11fa384f7e6ca248cL243) ?

Reply:
I mentioned in the PR desc that I omitted it intentionally cause it just seems pointless to me.

Reply:
Then, I think, you can remove `@/lib/highlight` helper - as I remember, it's the only place where it's used (more lines to delete! :smile:)

Reply:
_(but anyway check before removing - I may be wrong with "the only place")_

Reply:
Looks safe. Removed.